tms/watcher/watcher.py

```diff
@@ -0,0 +1,472 @@
+            got_new_events = True
+            # update
+            try:
+                cluster_infos[job_event.cluster].update_from_event(job_event)
+            except KeyError:
+                LOGGER.warning(
+                    f"Cluster found in job event log does not match any "
+                    f"known taskforce ({job_event.cluster}), skipping it"
+                )
+                continue
+            except UnknownJobEvent as e:
+                LOGGER.debug(f"error: {e}")
+            # check times
+            if interval_timer.has_been_x_seconds():
+                break
+
+        # endgame check
+        if (not got_new_events) and all(c.seen_in_jel for c in cluster_infos.values()):
+            if is_file_past_modification_expiry(jel_fpath):
+                # case: file has not been updated and it's old
+                jel_fpath.unlink()  # delete file
+                LOGGER.warning(f"Deleted job log file {jel_fpath}")
+                return
+            else:
+                # case: file has not been updated but need to wait longer
+                continue
+
+        LOGGER.info("Done reading events for now")
+        # TODO - remove
+        LOGGER.debug(
+            pprint.pformat({k: v._jobs for k, v in cluster_infos.items()}, indent=4)
+        )
+
+        patch_body: dict[str, Any] = {
+            _ALL_TOP_ERRORS_KEY: {},
+            _ALL_COMP_STAT_KEY: {},
+        }
+
+        # aggregate
+        # NOTE: We unfortunately cannot reduce the data after aggregating.
+        #  Once we aggregate we lose job-level granularity, which is
+        #  needed for replacing/updating individual jobs' status(es).
+        #  Alternatively, we could re-parse the entire job log every time.
+        for cid, info in cluster_infos.items():
+            try:
+                LOGGER.info(
+                    f"Getting top task errors {info.taskforce_uuid=} / {cid=}..."
+                )
+                patch_body[_ALL_TOP_ERRORS_KEY][
+                    info.taskforce_uuid
+                ] = info.get_top_task_errors()
+            except NoUpdateException:
+                pass
+            try:
+                LOGGER.info(
+                    f"Aggregating compound statuses {info.taskforce_uuid=} / {cid=}..."
+                )
+                patch_body[_ALL_COMP_STAT_KEY][
+                    info.taskforce_uuid
+                ] = info.aggregate_compound_statuses()
+            except NoUpdateException:
+                pass
+
+        # send -- one big update that way it can't intermittently fail
+        # (it's okay to send an empty sub-dict, but all empties is pointless)
+        if any(patch_body[k] for k in [_ALL_TOP_ERRORS_KEY, _ALL_COMP_STAT_KEY]):
+            LOGGER.info("Sending updates to EWMS...")
+            await ewms_rc.request(
+                "PATCH",
+                "/tms/taskforces/many",
+                patch_body,
```

Comment:
This feels more like a POST, since we're not actually directly updating a single resource, but running an action on the server given our input json.  Could probably also use a better URL name.

Reply:
updated: [`c447f5b` (#3)](https://github.com/Observation-Management-Service/ewms-task-management-service/pull/3/commits/c447f5b80ca5a592349d4201c05dac0ee2615ced)